File: presentation/src/main/java/com/ahn/presentation/ui/screen/exchange/ExchangeViewModel.kt
package com.ahn.presentation.ui.screen.exchange

import android.util.Log
import androidx.lifecycle.ViewModel
import com.ahn.domain.currency.model.CurrencyInfo
import com.ahn.domain.currency.model.ExchangeCurrencySelection
import com.ahn.domain.currency.usecase.CurrencySelectionUseCases
import com.ahn.domain.exchange.usecase.ExchangeUseCases
import com.ahn.domain.favorite.usecase.FavoriteUseCases
import com.ahn.presentation.R
import com.ahn.presentation.util.UiText
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.CancellationException
import org.orbitmvi.orbit.ContainerHost
import org.orbitmvi.orbit.syntax.Syntax
import org.orbitmvi.orbit.viewmodel.container
import javax.inject.Inject

@HiltViewModel
class ExchangeViewModel @Inject constructor(
    private val exchangeUseCases: ExchangeUseCases,
    private val favoriteUseCases: FavoriteUseCases,
    private val currencySelectionUseCases: CurrencySelectionUseCases,
) : ViewModel(), ContainerHost<ExchangeContract.State, ExchangeContract.SideEffect> {

    override val container = container(
        initialState = ExchangeContract.State(),
    ) {
        performLoadCurrencies()
        observeFavorites()
    }

    fun processIntent(intent: ExchangeContract.Intent) {
        when (intent) {
            is ExchangeContract.Intent.UpdateFromAmount -> handleUpdateFromAmount(intent.amount)
            is ExchangeContract.Intent.SelectFromCurrency -> handleSelectFromCurrency(intent.currency)
            is ExchangeContract.Intent.SelectToCurrency -> handleSelectToCurrency(intent.currency)
            is ExchangeContract.Intent.ToggleFavorite -> handleToggleFavorite(intent.currencyCode)
            is ExchangeContract.Intent.SwapCurrencies -> intent { performSwapCurrencies() }
            is ExchangeContract.Intent.LoadCurrencies -> intent { performLoadCurrencies() }
        }
    }

    private fun handleUpdateFromAmount(amount: String) = intent {
        if (amount.isEmpty() || amount.matches(Regex("^\\d*\\.?\\d*$"))) {
            reduce {
                state.copy(
                    fromAmount = amount,
                    toAmount = exchangeUseCases.exchangeAmount(amount, state.exchangeRate)
                )
            }
        }
    }

    private fun handleSelectFromCurrency(currency: CurrencyInfo) = intent {
        if (currency == state.toCurrency) {
            performSwapCurrencies()
        } else {
            saveExchangeSelectionBestEffort {
                currencySelectionUseCases.saveExchangeFromCurrency(currency.code)
            }

            reduce { state.copy(fromCurrency = currency) }
            performFetchExchangeRate()
        }
    }

    private fun handleSelectToCurrency(currency: CurrencyInfo) = intent {
        if (currency == state.fromCurrency) {
            performSwapCurrencies()
        } else {
            saveExchangeSelectionBestEffort {
                currencySelectionUseCases.saveExchangeToCurrency(currency.code)
            }

            reduce { state.copy(toCurrency = currency) }
            performFetchExchangeRate()
        }
    }

    private fun observeFavorites() = intent {
        favoriteUseCases.getFavoriteCurrencies().collect { codes ->
            reduce { state.copy(favoriteCurrencyCodes = codes) }
        }
    }

    private fun handleToggleFavorite(currencyCode: String) = intent {
        val wasFavorite = currencyCode in state.favoriteCurrencyCodes

        runCatching {
            favoriteUseCases.toggleFavoriteCurrency(currencyCode)
        }.onSuccess {
            postSideEffect(
                ExchangeContract.SideEffect.ShowSnackBar(
                    UiText.StringResource(
                        if (wasFavorite) R.string.favorite_removed else R.string.favorite_added
                    )
                )
            )
        }.onFailure { e ->
            if (e is CancellationException) throw e
            postSideEffect(
                ExchangeContract.SideEffect.ShowSnackBar(
                    UiText.StringResource(R.string.favorite_change_failed)
                )
            )
        }
    }

    private suspend fun Syntax<ExchangeContract.State, ExchangeContract.SideEffect>.performLoadCurrencies() {
        try {
            reduce { state.copy(isLoading = true) }

            val currencies = exchangeUseCases.getSupportedCurrencies()
            val krw = currencies.find { it.code == "KRW" }
            val usd = currencies.find { it.code == "USD" }
            val savedSelection = getSavedExchangeSelectionOrNull()

            val preservedFrom = state.fromCurrency?.let { current ->
                currencies.find { it.code == current.code }
            }

            val fromCurrency = preservedFrom
                ?: currencies.find { it.code == savedSelection?.fromCode }
                ?: usd
                ?: currencies.firstOrNull()
                ?: state.fromCurrency

            val preservedTo = state.toCurrency?.let { current ->
                currencies.find { it.code == current.code }
            }

            val toCurrency = preservedTo
                ?: currencies.find { it.code == savedSelection?.toCode && it.code != fromCurrency?.code }
                ?: krw?.takeIf { it.code != fromCurrency?.code }
                ?: currencies.firstOrNull { it.code != fromCurrency?.code }
                ?: state.toCurrency

            reduce {
                state.copy(
                    availableCurrencies = currencies,
                    fromCurrency = fromCurrency,
                    toCurrency = toCurrency,
                    isLoading = false
                )
            }

            if (state.fromCurrency != null && state.toCurrency != null) {
                performFetchExchangeRate()
            }
        } catch (e: Exception) {
            if (e is CancellationException) throw e
            reduce { state.copy(isLoading = false) }
            postSideEffect(
                ExchangeContract.SideEffect.ShowSnackBar(
                    UiText.StringResource(
                        R.string.load_currency_list_failed,
                        listOf(e.message.orEmpty()),
                    )
                )
            )
        }
    }

    private suspend fun Syntax<ExchangeContract.State, ExchangeContract.SideEffect>.performSwapCurrencies() {
        val newFrom = state.toCurrency ?: return
        val newTo = state.fromCurrency ?: return
        val newFromAmount = state.toAmount.ifEmpty { "1" }

        reduce {
            state.copy(
                fromCurrency = newFrom,
                toCurrency = newTo,
                fromAmount = newFromAmount
            )
        }

        saveExchangeSelectionBestEffort {
            currencySelectionUseCases.saveExchangeSelection(
                fromCode = newFrom.code,
                toCode = newTo.code,
            )
        }

        performFetchExchangeRate()
    }

    private suspend fun getSavedExchangeSelectionOrNull(): ExchangeCurrencySelection? {
        return try {
            currencySelectionUseCases.getExchangeSelection()
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            Log.w("ExchangeViewModel", "Failed to load saved exchange currency selection.", e)
            null
        }
    }

    private suspend fun saveExchangeSelectionBestEffort(action: suspend () -> Unit) {
        try {
            action()
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            Log.w("ExchangeViewModel", "Failed to save exchange currency selection.", e)
        }
    }

    private suspend fun Syntax<ExchangeContract.State, ExchangeContract.SideEffect>.performFetchExchangeRate() {
        val fromCurrency = state.fromCurrency ?: return
        val toCurrency = state.toCurrency ?: return

        val requestedFrom = fromCurrency.code
        val requestedTo = toCurrency.code

        try {
            reduce { state.copy(isLoading = true) }

            // 여기서 일시 중단(suspend)되어 API 응답을 기다림 (동시성 꼬임 방지)
            val rate = exchangeUseCases.getExchangeRate(
                from = requestedFrom,
                to = requestedTo,
            )
            val rateDate = exchangeUseCases.getLatestRateDate()

            // 응답이 돌아왔을 때 상태가 이미 바뀌었으면 적용하지 않음
            if (state.fromCurrency?.code != requestedFrom ||
                state.toCurrency?.code != requestedTo) return

            reduce {
                state.copy(
                    exchangeRate = rate,
                    exchangeRateDate = rateDate,
                    isLoading = false,
                    toAmount = exchangeUseCases.exchangeAmount(state.fromAmount, rate)
                )
            }
        } catch (e: Exception) {
            if (e is CancellationException) throw e
            reduce { state.copy(isLoading = false) }
            postSideEffect(
                ExchangeContract.SideEffect.ShowSnackBar(
                    UiText.StringResource(
                        R.string.load_exchange_rate_failed,
                        listOf(e.message.orEmpty()),
                    )
                )
            )
        }
    }
}

File: ios/Packages/Presentation/Sources/Presentation/Screen/Exchange/ExchangeViewModel.swift
import Domain
import Foundation
import Observation

/// 환율 화면 ViewModel (Android `ExchangeViewModel` 대응)
///
/// Android의 activity-scoped 공유와 같게, 앱에서 하나만 만들어
/// Exchange·Favorite·Setting 화면이 함께 쓴다.
@MainActor
@Observable
public final class ExchangeViewModel {
    public private(set) var state = ExchangeState()

    @ObservationIgnored private let exchangeUseCases: ExchangeUseCases
    @ObservationIgnored private let favoriteUseCases: FavoriteUseCases
    @ObservationIgnored private let currencySelectionUseCases: CurrencySelectionUseCases
    @ObservationIgnored private let sideEffectBus = SideEffectBus<ExchangeSideEffect>()
    @ObservationIgnored private var observeFavoritesTask: Task<Void, Never>?

    public init(
        exchangeUseCases: ExchangeUseCases,
        favoriteUseCases: FavoriteUseCases,
        currencySelectionUseCases: CurrencySelectionUseCases,
    ) {
        self.exchangeUseCases = exchangeUseCases
        self.favoriteUseCases = favoriteUseCases
        self.currencySelectionUseCases = currencySelectionUseCases

        Task { await performLoadCurrencies() }
        observeFavorites()
    }

    deinit {
        observeFavoritesTask?.cancel()
    }

    public func sideEffects() -> AsyncStream<ExchangeSideEffect> {
        sideEffectBus.stream()
    }

    public func send(_ intent: ExchangeIntent) {
        switch intent {
        case let .updateFromAmount(amount): handleUpdateFromAmount(amount)
        case let .selectFromCurrency(currency): Task { await handleSelectFromCurrency(currency) }
        case let .selectToCurrency(currency): Task { await handleSelectToCurrency(currency) }
        case let .toggleFavorite(currencyCode): Task { await handleToggleFavorite(currencyCode) }
        case .swapCurrencies: Task { await performSwapCurrencies() }
        case .loadCurrencies: Task { await performLoadCurrencies() }
        case .refreshExchangeRates: Task { await performRefreshExchangeRates() }
        }
    }

    private func handleUpdateFromAmount(_ amount: String) {
        guard isValidAmountInput(amount) else { return }

        state.fromAmount = amount
        state.toAmount = exchangeUseCases.exchangeAmount(fromAmount: amount, rate: state.exchangeRate)
    }

    /// 숫자와 소수점 하나만 허용한다 (Android `^\d*\.?\d*$` 대응)
    private func isValidAmountInput(_ amount: String) -> Bool {
        amount.allSatisfy { $0.isNumber || $0 == "." } && amount.count(where: { $0 == "." }) <= 1
    }

    private func handleSelectFromCurrency(_ currency: CurrencyInfo) async {
        if currency == state.toCurrency {
            await performSwapCurrencies()
        } else {
            try? await currencySelectionUseCases.saveExchangeFromCurrency(currency.code)

            state.fromCurrency = currency
            await performFetchExchangeRate()
        }
    }

    private func handleSelectToCurrency(_ currency: CurrencyInfo) async {
        if currency == state.fromCurrency {
            await performSwapCurrencies()
        } else {
            try? await currencySelectionUseCases.saveExchangeToCurrency(currency.code)

            state.toCurrency = currency
            await performFetchExchangeRate()
        }
    }

    private func observeFavorites() {
        let favoritesStream = favoriteUseCases.getFavoriteCurrencies()

        observeFavoritesTask = Task { [weak self] in
            for await codes in favoritesStream {
                guard let self else { return }

                state.favoriteCurrencyCodes = codes
            }
        }
    }

    private func handleToggleFavorite(_ currencyCode: String) async {
        let wasFavorite = state.favoriteCurrencyCodes.contains(currencyCode)

        do {
            try await favoriteUseCases.toggleFavoriteCurrency(currencyCode)
            sideEffectBus.send(
                .showSnackbar(message: wasFavorite ? L("즐겨찾기가 해제되었습니다.") : L("즐겨찾기에 추가되었습니다.")),
            )
        } catch {
            sideEffectBus.send(.showSnackbar(message: L("즐겨찾기 변경에 실패했습니다.")))
        }
    }

    private func performLoadCurrencies() async {
        do {
            state.isLoading = true

            let currencies = try await exchangeUseCases.getSupportedCurrencies()
            let savedSelection = try? await currencySelectionUseCases.getExchangeSelection()
            let selected = resolveExchangeCurrencies(
                currencies: currencies,
                savedSelection: savedSelection,
                currentFromCurrency: state.fromCurrency,
                currentToCurrency: state.toCurrency,
            )

            state.availableCurrencies = currencies
            state.fromCurrency = selected.from
            state.toCurrency = selected.to
            state.isLoading = false

            if selected.from != nil, selected.to != nil {
                await performFetchExchangeRate()
            }
        } catch {
            state.isLoading = false
            sideEffectBus.send(.showSnackbar(message: error.exchangeRateErrorMessage))
        }
    }

    private func resolveExchangeCurrencies(
        currencies: [CurrencyInfo],
        savedSelection: ExchangeCurrencySelection?,
        currentFromCurrency: CurrencyInfo?,
        currentToCurrency: CurrencyInfo?,
    ) -> (from: CurrencyInfo?, to: CurrencyInfo?) {
        let krw = currencies.first { $0.code == "KRW" }
        let usd = currencies.first { $0.code == "USD" }

        let fromCurrency = currencies.first { $0.code == currentFromCurrency?.code }
            ?? currencies.first { $0.code == savedSelection?.fromCode }
            ?? usd
            ?? currencies.first
            ?? currentFromCurrency

        let fromCode = fromCurrency?.code
        let toCurrency = currencies.first { $0.code == currentToCurrency?.code }
            ?? currencies.first { $0.code == savedSelection?.toCode && $0.code != fromCode }
            ?? krw.flatMap { $0.code != fromCode ? $0 : nil }
            ?? currencies.first { $0.code != fromCode }
            ?? currentToCurrency

        return (fromCurrency, toCurrency)
    }

    private func performSwapCurrencies() async {
        guard let newFrom = state.toCurrency, let newTo = state.fromCurrency else { return }

        state.fromCurrency = newFrom
        state.toCurrency = newTo
        state.fromAmount = state.toAmount.isEmpty ? "1" : state.toAmount

        try? await currencySelectionUseCases.saveExchangeSelection(
            fromCode: newFrom.code,
            toCode: newTo.code,
        )

        await performFetchExchangeRate()
    }

    private func performFetchExchangeRate() async {
        guard let fromCurrency = state.fromCurrency, let toCurrency = state.toCurrency else { return }

        let requestedFrom = fromCurrency.code
        let requestedTo = toCurrency.code

        do {
            state.isLoading = true

            // 여기서 일시 중단(suspend)되어 응답을 기다림 (동시성 꼬임 방지)
            let rate = try await exchangeUseCases.getExchangeRate(from: requestedFrom, to: requestedTo)
            let rateDate = try await exchangeUseCases.getLatestRateDate()
            let fetchedAt = try await exchangeUseCases.getLatestFetchedAt()

            // 응답이 돌아왔을 때 상태가 이미 바뀌었으면 적용하지 않음
            guard state.fromCurrency?.code == requestedFrom, state.toCurrency?.code == requestedTo else {
                return
            }

            state.exchangeRate = rate
            state.exchangeRateDate = rateDate
            state.exchangeRateFetchedAt = fetchedAt
            state.isLoading = false
            state.toAmount = exchangeUseCases.exchangeAmount(fromAmount: state.fromAmount, rate: rate)
        } catch {
            state.isLoading = false
            sideEffectBus.send(.showSnackbar(message: error.exchangeRateErrorMessage))
        }
    }

    private func performRefreshExchangeRates() async {
        do {
            state.isLoading = true

            try await exchangeUseCases.refreshExchangeRates()

            let currencies = try await exchangeUseCases.getSupportedCurrencies()
            let savedSelection = try? await currencySelectionUseCases.getExchangeSelection()
            let selected = resolveExchangeCurrencies(
                currencies: currencies,
                savedSelection: savedSelection,
                currentFromCurrency: state.fromCurrency,
                currentToCurrency: state.toCurrency,
            )
            let rate = try await exchangeRateOrDefault(fromCurrency: selected.from, toCurrency: selected.to)
            let rateDate = try await exchangeUseCases.getLatestRateDate()
            let fetchedAt = try await exchangeUseCases.getLatestFetchedAt()

            state.availableCurrencies = currencies
            state.fromCurrency = selected.from
            state.toCurrency = selected.to
            state.exchangeRate = rate
            state.exchangeRateDate = rateDate
            state.exchangeRateFetchedAt = fetchedAt
            state.isLoading = false
            state.toAmount = exchangeUseCases.exchangeAmount(fromAmount: state.fromAmount, rate: rate)

            sideEffectBus.send(.showSnackbar(message: L("환율 정보를 새로고침했습니다.")))
        } catch {
            state.isLoading = false
            sideEffectBus.send(.showSnackbar(message: error.exchangeRateErrorMessage))
        }
    }

    private func exchangeRateOrDefault(
        fromCurrency: CurrencyInfo?,
        toCurrency: CurrencyInfo?,
    ) async throws -> Double {
        guard let fromCurrency, let toCurrency else { return 0.0 }
        if fromCurrency.code == toCurrency.code { return 1.0 }

        return try await exchangeUseCases.getExchangeRate(from: fromCurrency.code, to: toCurrency.code)
    }
}
